src/catalog/src/process_manager.rs

```diff
@@ -0,0 +1,181 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::{Arc, RwLock};
+
+use api::v1::frontend::{ListProcessRequest, ProcessInfo};
+use common_frontend::selector::{FrontendSelector, MetaClientSelector};
+use common_telemetry::{debug, info};
+use common_time::util::current_time_millis;
+use meta_client::MetaClientRef;
+use snafu::ResultExt;
+
+use crate::error;
+
+pub type ProcessManagerRef = Arc<ProcessManager>;
+
+pub struct ProcessManager {
+    server_addr: String,
+    next_id: AtomicU64,
+    catalogs: RwLock<HashMap<String, HashMap<u64, ProcessInfo>>>,
+    frontend_selector: Option<MetaClientSelector>,
+}
+
+impl ProcessManager {
+    /// Create a [ProcessManager] instance with server address and kv client.
+    pub fn new(server_addr: String, meta_client: Option<MetaClientRef>) -> Self {
+        let frontend_selector = meta_client.map(MetaClientSelector::new);
+        Self {
+            server_addr,
+            next_id: Default::default(),
+            catalogs: Default::default(),
+            frontend_selector,
+        }
+    }
+}
+
+impl ProcessManager {
+    /// Registers a submitted query.
+    pub fn register_query(
+        self: &Arc<Self>,
+        catalog: String,
+        schemas: Vec<String>,
+        query: String,
+        client: String,
+    ) -> Ticket {
+        let id = self.next_id.fetch_add(1, Ordering::Relaxed);
```

Comment:
Due to historical reason, the process_id in postgres protocol is assigned to connection because there is no multiplex support in postgres protocol, so killing a query is killing a connection, which is backed by a process. That's why it's called Process. I think mysql works like this too.

I will suggest to add an optional `pid` argument for this function so when query is registered from postgres or mysql, we will use the pre-assigned pid, while if it's from http or other stateless interface, pid is generated per-query.

This will also require a new API from `ProcessManager` called like `next_pid()` and it has to be called upon connection of postgres and mysql.

Reply:
added in e516ae49e

Reply:
Looks good. I have opened https://github.com/GreptimeTeam/greptimedb/pull/6301/ for `Session` and `QueryContext`.